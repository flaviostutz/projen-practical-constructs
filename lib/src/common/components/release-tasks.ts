import { Component, Project } from 'projen';

import { expandMonotagCmd, NextTagOptions } from '../types/monotag';
import { WithRequired } from '../types/utils';
import { monotagCliArgs } from '../utils/monotag';

import { CommonTargets } from './common-target-type';

/**
 * Create tasks to support the basics of creating a software release based on git tags.
 * Uses monotag to calculate the next tag and release notes.
 *
 * Tasks:
 *   - release[:name]:current: Verifies if the current commit is already tagged with the latest calculated tag. If so, bumps files, saves version/tag/notes in output files/changelogs.
 *   - release[:name]: Will execute before -> next-tag -> generate -> [tag|push] -> after for this release group
 *   - release[:name]:before: executed before any other tasks in release. Placeholder for any pre-release related tasks
 *   - release[:name]:next-tag: Calculate the next version of the software, output tag and notes to console, write output files, bump files and append changelog. Supports complex release tagging in monorepos by using "npx monotag"
 *   - release[:name]:generate: executed after tag calculation and before git operations. Placeholder for custom tasks, such as doc generation, package build etc
 *   - release[:name]:git-tag: Tag the current commit with the calculated release tag on git repo (if action is 'tag')
 *   - release[:name]:git-push: Push the tagged commit to remote git (if action is 'push')
 *   - release[:name]:after: executed after all other tasks in release. Placeholder for any post-release related tasks
 */
export class ReleaseTasks extends Component {
  constructor(project: Project, opts?: ReleaseOptions) {
    super(project);

    const optsWithDefaults = getOptionsWithDefaults(opts);

    // add resources to gitignore
    if (optsWithDefaults.notesFile) {
      project.addGitIgnore(optsWithDefaults.notesFile);
    }
    if (optsWithDefaults.tagFile) {
      project.addGitIgnore(optsWithDefaults.tagFile);
    }
    if (optsWithDefaults.versionFile) {
      project.addGitIgnore(optsWithDefaults.versionFile);
    }

    const taskPrefix = `${CommonTargets.RELEASE}${optsWithDefaults.name ? `:${optsWithDefaults.name}` : ''}`;

    const releaseTask = project.addTask(`${taskPrefix}`, {
      description: `Release a new version by calculating next tag/version, generating changelogs, documentation, commiting, tagging and pushing these changes/tag to the repo.`,
    });

    const preTask = project.addTask(`${taskPrefix}:before`, {
      description: 'Executed before any release tasks. Placeholder for customizations',
    });
    releaseTask.spawn(preTask);

    const nextTagTask = project.addTask(`${taskPrefix}:next-tag`, {
      description:
        'Calculate next tag and version of the software and display on console. No git operations are performed, but depending on the configurations, changelog, version, tag an notes might be written to files. Supports complex release tagging in monorepos by using "npx monotag"',
      steps: [
        {
          exec: `${optsWithDefaults.monotagCmd} tag ${monotagCliArgs(optsWithDefaults)}`,
        },
      ],
    });
    releaseTask.spawn(nextTagTask);

    project.addTask(`${taskPrefix}:current`, {
      description:
        'Verifies if the current commit is already tagged with the latest calculated tag. If so, bumps files, saves version/tag/notes in output files/changelogs.',
      steps: [
        {
          exec: `${optsWithDefaults.monotagCmd} current ${monotagCliArgs(optsWithDefaults)}`,
        },
      ],
    });

    const generateTask = project.addTask(`${taskPrefix}:generate`, {
      description: 'Generates documentation, special files etc. Placeholder for customizations',
    });
    releaseTask.spawn(generateTask);

    if (optsWithDefaults.action === 'tag') {
      const gitTagTask = project.addTask('git-tag', {
        description: 'Commit all files and git tag with calculated release tag',
        steps: [
          {
            exec: `${optsWithDefaults.monotagCmd} tag-git ${monotagCliArgs(optsWithDefaults)}`,
          },
        ],
      });
      releaseTask.spawn(gitTagTask);
    }

    if (optsWithDefaults.action === 'push') {
      const gitPushTask = project.addTask(`${taskPrefix}:git-push`, {
        description:
          'Commit all files, git tag with calculated release tag and push it to remote git',
        steps: [
          {
            exec: `${optsWithDefaults.monotagCmd} tag-push ${monotagCliArgs(optsWithDefaults)}`,
          },
        ],
      });
      releaseTask.spawn(gitPushTask);
    }

    const postTask = project.addTask(`${taskPrefix}:after`, {
      description: 'Executed after all release tasks. Placeholder for customizations',
    });
    releaseTask.spawn(postTask);
  }
}

const getOptionsWithDefaults = (
  opts?: ReleaseOptions,
): WithRequired<ReleaseOptions, 'name' | 'action' | 'monotagCmd'> => {
  // check if action is valid
  if (opts?.action && !['console', 'tag', 'push'].includes(opts.action)) {
    throw new Error(
      `Invalid action: '${opts?.action}'. Valid actions are: 'console', 'tag', 'push'`,
    );
  }

  if (
    (opts?.action === 'push' || opts?.action === 'tag') &&
    (!opts?.gitEmail || !opts?.gitUsername)
  ) {
    throw new Error('gitEmail and gitUsername are required when action is "tag" or "push"');
  }

  return {
    name: '',
    action: 'console',
    monotagCmd: expandMonotagCmd(opts?.monotagCmd),
    ...opts,
  };
};

export interface ReleaseOptions extends NextTagOptions {
  /**
   * Action to be taken after calculating the next tag
   * Options:
   *  - console: Print calculated tag/notes to console
   *  - tag: Calculate tag/notes, commit and tag (git) resources
   *  - push: Calculate tag/notes, commit, tag (git) and push resources to remote git
   * @default 'console'
   */
  readonly action?: 'console' | 'tag' | 'push';
  /**
   * Name of this release group of tasks
   * Useful if you have multiple release tasks in the same project
   * with different configurations.
   * The release tasks will be named as "release:[name]:[task]"
   * @default ''
   */
  readonly name?: string;
}
