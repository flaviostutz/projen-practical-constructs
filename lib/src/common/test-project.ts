import { mkdtempSync } from 'node:fs';
import { tmpdir } from 'node:os';
// eslint-disable-next-line unicorn/import-style
import { join } from 'node:path';

import { Project } from 'projen';

export class TestProject extends Project {
  constructor() {
    const outdir = mkdtempSync(join(tmpdir(), 'test-'));
    super({ name: 'test', outdir });
  }
}
