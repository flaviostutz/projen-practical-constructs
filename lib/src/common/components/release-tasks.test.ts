/* eslint-disable no-new */
import { Testing } from 'projen';

import { TestProject } from '../test-project';

import { ReleaseTasks, ReleaseTasksOptions } from './release-tasks';

describe('ReleaseTasks', () => {
  it('synthesizes correctly with default options', () => {
    const project = new TestProject();
    new ReleaseTasks(project);
    const output = Testing.synth(project);
    // eslint-disable-next-line unicorn/no-null
    expect(output).toMatchSnapshot();
  });

  it('synthesizes correctly with custom options', () => {
    const project = new TestProject();
    const options: ReleaseTasksOptions = {
      action: 'push',
      name: 'custom',
      gitEmail: 'test@example.com',
      gitUsername: 'testuser',
    };
    new ReleaseTasks(project, options);
    const output = Testing.synth(project);
    expect(output).toMatchSnapshot();
  });

  it('adds options to monotag calls', () => {
    const project = new TestProject();
    const options: ReleaseTasksOptions = {
      action: 'push',
      name: 'custom',
      gitEmail: 'test@example.com',
      gitUsername: 'testuser',
      bumpAction: 'latest',
      preRelease: true,
    };
    new ReleaseTasks(project, options);
    const output = Testing.synth(project);
    const tasksObj = output['.projen/tasks.json'];
    expect(tasksObj).toMatchObject({
      tasks: {
        'release:custom:next-tag': {
          steps: [
            {
              exec: 'npx -y monotag@1.14.0 tag --bump-action="latest" --git-email="test@example.com" --git-username="testuser" --prerelease="true"',
            },
          ],
        },
      },
    });
  });

  it('throws error for invalid action', () => {
    const project = new TestProject();
    const options: ReleaseTasksOptions = {
      action: 'invalid' as 'console',
    };
    expect(() => new ReleaseTasks(project, options)).toThrow(
      'Invalid action: invalid. Valid actions are: console, tag, push',
    );
  });

  it('throws error when gitEmail and gitUsername are missing for tag action', () => {
    const project = new TestProject();
    const options: ReleaseTasksOptions = {
      action: 'tag',
    };
    expect(() => new ReleaseTasks(project, options)).toThrow(
      'gitEmail and gitUsername are required when action is "tag" or "push"',
    );
  });

  it('throws error when gitEmail and gitUsername are missing for push action', () => {
    const project = new TestProject();
    const options: ReleaseTasksOptions = {
      action: 'push',
    };
    expect(() => new ReleaseTasks(project, options)).toThrow(
      'gitEmail and gitUsername are required when action is "tag" or "push"',
    );
  });
});
