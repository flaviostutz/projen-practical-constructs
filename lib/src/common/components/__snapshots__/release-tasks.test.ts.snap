// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleaseTasks synthesizes correctly with custom options 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "release:custom": {
        "description": "Release a new version by calculating next tag/version, generating changelogs, documentation, commiting, tagging and pushing these changes/tag to the repo.",
        "name": "release:custom",
        "steps": [
          {
            "spawn": "release:custom:before",
          },
          {
            "spawn": "release:custom:next-tag",
          },
          {
            "spawn": "release:custom:generate",
          },
          {
            "spawn": "release:custom:git-push",
          },
          {
            "spawn": "release:custom:after",
          },
        ],
      },
      "release:custom:after": {
        "description": "Executed after all release tasks. Placeholder for customizations",
        "name": "release:custom:after",
      },
      "release:custom:before": {
        "description": "Executed before any release tasks. Placeholder for customizations",
        "name": "release:custom:before",
      },
      "release:custom:current": {
        "description": "Verifies if the current commit is already tagged with the latest calculated tag. If so, bumps files, saves version/tag/notes in output files/changelogs.",
        "name": "release:custom:current",
        "steps": [
          {
            "exec": "npx -y monotag@1.23.0 current --git-email="test@example.com" --git-username="testuser"",
          },
        ],
      },
      "release:custom:generate": {
        "description": "Generates documentation, special files etc. Placeholder for customizations",
        "name": "release:custom:generate",
      },
      "release:custom:git-push": {
        "description": "Commit all files, git tag with calculated release tag and push it to remote git",
        "name": "release:custom:git-push",
        "steps": [
          {
            "exec": "npx -y monotag@1.23.0 tag-push --git-email="test@example.com" --git-username="testuser"",
          },
        ],
      },
      "release:custom:next-tag": {
        "description": "Calculate next tag and version of the software and display on console. No git operations are performed, but depending on the configurations, changelog, version, tag an notes might be written to files. Supports complex release tagging in monorepos by using "npx monotag"",
        "name": "release:custom:next-tag",
        "steps": [
          {
            "exec": "npx -y monotag@1.23.0 tag --git-email="test@example.com" --git-username="testuser"",
          },
        ],
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
}
`;

exports[`ReleaseTasks synthesizes correctly with default options 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "release": {
        "description": "Release a new version by calculating next tag/version, generating changelogs, documentation, commiting, tagging and pushing these changes/tag to the repo.",
        "name": "release",
        "steps": [
          {
            "spawn": "release:before",
          },
          {
            "spawn": "release:next-tag",
          },
          {
            "spawn": "release:generate",
          },
          {
            "spawn": "release:after",
          },
        ],
      },
      "release:after": {
        "description": "Executed after all release tasks. Placeholder for customizations",
        "name": "release:after",
      },
      "release:before": {
        "description": "Executed before any release tasks. Placeholder for customizations",
        "name": "release:before",
      },
      "release:current": {
        "description": "Verifies if the current commit is already tagged with the latest calculated tag. If so, bumps files, saves version/tag/notes in output files/changelogs.",
        "name": "release:current",
        "steps": [
          {
            "exec": "npx -y monotag@1.23.0 current ",
          },
        ],
      },
      "release:generate": {
        "description": "Generates documentation, special files etc. Placeholder for customizations",
        "name": "release:generate",
      },
      "release:next-tag": {
        "description": "Calculate next tag and version of the software and display on console. No git operations are performed, but depending on the configurations, changelog, version, tag an notes might be written to files. Supports complex release tagging in monorepos by using "npx monotag"",
        "name": "release:next-tag",
        "steps": [
          {
            "exec": "npx -y monotag@1.23.0 tag ",
          },
        ],
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
}
`;
